### ISSUE TITLE
Unnecessary referencing may cause infinite recursion in compilation

### ISSUE THREAD

**Issue by momvart** [BurntSushi/aho-corasick#149]
These two default implementations are passing a `&&Self` instead of `&Self` to the target function in place of `&A` where `A: Automaton`. This is unnecessary as `Self` is implementing `Automaton` so we can pass `self` directly.

https://github.com/BurntSushi/aho-corasick/blob/948d2e1f8e4b6b0aff13075176922e158c8bed46/src/automaton.rs#L358
and 
https://github.com/BurntSushi/aho-corasick/blob/948d2e1f8e4b6b0aff13075176922e158c8bed46/src/automaton.rs#L372

Although this is not a problem by itself, it can possibly cause an infinite loop when converting an `Automaton` to a dynamic object (as theoretically in each call a new type is evaluated for the implementation of `Automaton`). (Unfortunately I cannot regenerate how I managed to cause the compile error, but it's possible!).

**Comment by momvart** [comment]
BTW, I'm wondering why such implementation (`impl<'a, A: Automaton + ?Sized> Automaton for &'a A`) should exist. I'll open a PR to remove it if you don't mind.

**Comment by BurntSushi** [comment]
That trait impl can't be removed. It's a breaking change.

I don't remember the precise reason it exists, other than to make `&A` implement `Automaton`. My recollection is that it was also related to avoiding monomorphization creating extra copies of some functions. But I am not confident in my methodology here.

As for the infinite loop, I think I need an MRE to determine whether the fault lies with this crate or not.

**Comment by BurntSushi** [comment]
Right. I absolutely agree that you can get an infinite loop during compilation if the trait implemented incorrectly.

But all righty, I'm going to close this as a non-issue for now. Thanks for looking into this!
